django/db/backends/postgresql/base.py

```diff
@@ -179,6 +197,45 @@ class DatabaseWrapper(BaseDatabaseWrapper):
     ops_class = DatabaseOperations
     # PostgreSQL backend-specific attributes.
     _named_cursor_idx = 0
+    _connection_pools = {}
```

Comment:
Another regular safety hazard with mutable global resources aside from thread safety is "fork safety" - what happens with the resource in the child process after a fork. AFAICT psycopg pool is not "fork safe" (https://github.com/psycopg/psycopg/issues/651#issuecomment-1738659965). There are of course many WSGI servers using fork. What I'm not sure about is whether Django tries to be compatible with this scenario:

- Process opens a connection
- Process forks
- Child opens a connection

or if it requires the master process to not do any DB connections before forking.

(With existing connections/persistent connections code this is OK because the connections are stored in a thread local which gets reset on fork.)

Reply:
Is it really okay with persistent connections? After fork if the thread local gets reset wouldn't that mean that the connection is garbage collected which could possibly close the tcp connection? I don't think anything in Django is written with fork safety in mind. While WSGI servers use fork it generally only works if the fork early enough I think (preferably before loading the WSGI app)

Reply:
> Is it really okay with persistent connections? After fork if the thread local gets reset wouldn't that mean that the connection is garbage collected which could possibly close the tcp connection?

libpq (the underlying C lib used by psycopg) is using `O_CLOEXEC` (since [a long time](https://github.com/postgres/postgres/commit/7627b91cd5d) ago at least) so the socket isn't inherited in fork.

>  I don't think anything in Django is written with fork safety in mind. While WSGI servers use fork it generally only works if the fork early enough I think (preferably before loading the WSGI app)

I myself really only use thread and gevent workers so I don't have much operation experience with how process workers work. I found the following description in uWSGI docs ([link](https://uwsgi-docs.readthedocs.io/en/latest/articles/TheArtOfGracefulReloading.html#preforking-vs-lazy-apps-vs-lazy)):

"By default uWSGI loads the whole application in the first process and after the app is loaded it does fork() itself multiple times."

My guess is that it at least means it imports the wsgi entry point, which in turns does `django.setup()`. Is it guaranteed that `django.setup()` doesn't cause any queries to run? (The warning [here](https://docs.djangoproject.com/en/5.0/ref/applications/#django.apps.AppConfig.ready) certainly advises against it...)

Reply:
> Is it guaranteed that django.setup() doesn't cause any queries to run? (The warning [here](https://docs.djangoproject.com/en/5.0/ref/applications/#django.apps.AppConfig.ready) certainly advises against it...)

Since Django 5.0, this raises a Runtimewarning. So not guaranteed not to happen but actively discouraged.

Reply:
This is good enough for me!